### DISCUSSION TITLE
Raspberry Pi Compute Module 3 and I2C

### DISCUSSION THREAD

**Discussion by rengglian** [Pi4J/pi4j#313] [Q&A]
Hi all, 

I have following raspbian installed on my compute module 3:

> cat /etc/os-release
> PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
> NAME="Raspbian GNU/Linux"
> VERSION_ID="11"
> VERSION="11 (bullseye)"
> VERSION_CODENAME=bullseye
> ID=raspbian
> ID_LIKE=debian
> HOME_URL="http://www.raspbian.org/"
> SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
> BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"

with java:

> java -version
> openjdk version "17.0.9" 2023-10-17
> OpenJDK Runtime Environment (build 17.0.9+9-Raspbian-1deb11u1rpt1)
> OpenJDK Client VM (build 17.0.9+9-Raspbian-1deb11u1rpt1, mixed mode, emulated-client)

with javafx:
> javafx.version=11.0.10-internal
> javafx.runtime.version=11.0.10-internal+0-2021-02-03-151839
> javafx.runtime.build=0

with pigpio:
> apt list pigpio -a
> Listing... Done
> pigpio/oldstable,now 1.79-1+rpt1 armhf [installed]
> pigpio/oldstable 1.78-1+rpi1 armhf

In my I2cController class I create my device (which should control the light intensity of some LEDs):
```
private static I2C i2cDevice;
...
I2CProvider i2CProvider = pi4j.provider("pigpio-i2c");
I2CConfig i2cConfig = I2C.newConfigBuilder(pi4j)
       .id("ltc2635")
       .name("LTC2635")
       .bus(I2C_BUS)
       .device(slaveAddr)
       .build();
i2cDevice = i2CProvider.create(i2cConfig);
```

And upon user interaction on the screen, I try to write to my device:
```
public static void processLightLevel(byte lightReg, byte lightLevel)
		throws IOException {

	byte[] command = new byte[] {lightReg, lightLevel, 0x00 };
	i2cDevice.write(command);
}
```

GPIOS work, bu nothing on I2C happens.

Only when I close my application the light settings change.
It seems there is some kind of flush missing, or clock issue. any suggestions?

Slave address is 0x10
Light register is 0x30 or 0x31

tests with i2c-tools work:
`
i2cset -y 1 0x10 0x30 0xA0 0x00 i
`

its done with maven:
```
<pi4j.version>2.4.0</pi4j.version>

<dependency>
	<groupId>com.pi4j</groupId>
	<artifactId>pi4j-core</artifactId>
	<version>${pi4j.version}</version>
</dependency>  

<dependency>
	<groupId>com.pi4j</groupId>
	<artifactId>pi4j-plugin-raspberrypi</artifactId>
	<version>${pi4j.version}</version>
</dependency>               

<dependency>
	<groupId>com.pi4j</groupId>
	<artifactId>pi4j-plugin-pigpio</artifactId>
	<version>${pi4j.version}</version>
</dependency>  
```

**Comment by taartspi** [comment]
Seeing little of your code i will only be able to ask questions.  Does your chip have the ldac pin ?   You said ending the app effects the output.  If u used the ldac and drove it high and did not drive low after sending the i2c data AND was configured to go low when the app exists, you could see this behavior

**Comment by rengglian** [comment]
Good question about the ldac pin, I have to check the HW/schematics when I'm back in the office. 
There is at least no sign of it in the code base I obtained. 
I upgraded the project from pi4j V1, where it worked (wiringpi, and java 1.8 / 8). 

Here the code and pom.xml snippets:
```
	public static void processLightLevel(int devicceReg, byte lightReg, byte lightLevel)
			throws UnsupportedBusNumberException, IOException {
		I2CBus i2c = I2CFactory.getInstance(I2CBus.BUS_1);
		I2CDevice device = i2c.getDevice(devicceReg);

		// Changing the Lightintensity on Hardware
		byte[] bytes = new byte[] { lightLevel, 0x00 };
		device.write(lightReg, bytes);
	}
```

```
            <pi4j.version>1.2</pi4j.version>

            <!-- https://mvnrepository.com/artifact/com.pi4j/pi4j-core -->
            <dependency>
                <groupId>com.pi4j</groupId>
                <artifactId>pi4j-core</artifactId>
                <version>${pi4j.version}</version>
            </dependency>

            <!-- https://mvnrepository.com/artifact/com.pi4j/pi4j-device -->
            <dependency>
                <groupId>com.pi4j</groupId>
                <artifactId>pi4j-device</artifactId>
                <version>${pi4j.version}</version>
            </dependency>

            <!-- https://mvnrepository.com/artifact/com.pi4j/pi4j-gpio-extension -->
            <dependency>
                <groupId>com.pi4j</groupId>
                <artifactId>pi4j-gpio-extension</artifactId>
                <version>${pi4j.version}</version>
            </dependency>  
```
